File: ios/Plugin/ADEUMMobileCapacitorPluginPlugin.swift
import Foundation
import Capacitor
import ADEUMInstrumentation



/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(ADEUMMobileCapacitorPluginPlugin)
public class ADEUMMobileCapacitorPluginPlugin: CAPPlugin {
    private let implementation = ADEUMMobileCapacitorPlugin()
    private var config = ADEumAgentConfiguration()
    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve([
            "value": implementation.echo(value)
        ])
    }
    @objc func startTimer(_ call: CAPPluginCall) {
        let name = call.getString("name") ?? "generic_timer"
        implementation.startTimerWithName(name: name)
        call.resolve()
    }
    @objc func stopTimer(_ call: CAPPluginCall) {
        let name = call.getString("name") ?? "generic_timer"
        implementation.stopTimerWithName(name: name)
        call.resolve()
    }
    @objc func reportMetricWithName(_ call: CAPPluginCall) {
        let name = call.getString("name") ?? "generic_metric"
        let value = Int64(call.getInt("value", 0))
        implementation.reportMetricWithName(name: name, value: value)
        call.resolve()
    }
    @objc func leaveBreadcrumb(_ call: CAPPluginCall) {
        let name = call.getString("name") ?? nil //will not record nil or empty string
        implementation.leaveBreadcrumb(name: name)
        call.resolve()
    }
    @objc func setUserData(_ call: CAPPluginCall) {
        let key = call.getString("key") ?? nil
        let value = call.getString("value") ?? ""
        //ignore if key is nil
        if key == nil {
            call.resolve()
            return
        }
        
        implementation.setUserData(key: key!, value: value)
        call.resolve()
    }
    @objc func removeUserData(_ call: CAPPluginCall) {
        let key = call.getString("key") ?? nil
        //ignore if key is nil
        if key == nil {
            call.resolve()
            return
        }
        
        implementation.removeUserData(key: key!)
        call.resolve()
    }
    @objc func takeScreenShot(_ call: CAPPluginCall) {
        implementation.takeScreenshot()
        call.resolve()
    }
    @objc func beginCall(_ call: CAPPluginCall) {
        let className = call.getString("className") ?? nil
        let methodName = call.getString("methodName") ?? nil
        let withArguments = call.getArray("withArguments", [])
        
        if className == nil || methodName == nil {
            //do nothing and return nothing
            call.resolve()
            return
        }
        call.resolve([
            "call_tracker": implementation.beginCall(className: className!, methodName: methodName!, withArguments: withArguments) as String
        ])
        return
    }
    @objc func endCall(_ call: CAPPluginCall) {
        let tracker = call.getString("call_tracker")
        if tracker != nil {
            implementation.endCall(tracker_key: tracker!)
        }
        call.resolve()
    }
    @objc func beginHttpRequest(_ call: CAPPluginCall) {
        let url_string = call.getString("url") ?? nil
        if url_string == nil {
            //do nothing and return nothing
            call.resolve()
            return
        }
        let url = URL.init(string:url_string!)
        let http_tracker = implementation.beginHttpRequest(url: url!)
        call.resolve([
            "http_tracker": http_tracker
        ])
        return
    }
    @objc func reportDone(_ call: CAPPluginCall) {
        let tracker = call.getString("http_tracker")
        if tracker != nil {
            implementation.reportDone(tracker_key: tracker!)
        }
        call.resolve()
    }
    
    @objc func withResponseCode(_ call: CAPPluginCall) {
        let tracker = call.getString("http_tracker")
        let statusCode = NSNumber.init(nonretainedObject: call.getInt("status_code"))
        if tracker != nil {
            implementation.withResponseCode(tracker_key: tracker!, statusCode: statusCode)
        }
        call.resolve()
    }
    
    @objc func withResponseContentLength(_ call: CAPPluginCall) {
        let tracker = call.getString("http_tracker")
        let content_length = NSNumber.init(nonretainedObject: call.getInt("content_length"))
        if tracker != nil {
            implementation.withResponseContentLength(tracker_key: tracker!, responseContentLength: content_length)
        }
        call.resolve()
    }
    
    @objc func withRequestContentLength(_ call: CAPPluginCall) {
        let tracker = call.getString("http_tracker")
        let content_length = NSNumber.init(nonretainedObject: call.getInt("content_length"))
        if tracker != nil {
            implementation.withRequestContentLength(tracker_key: tracker!, requestContentLength: content_length)
        }
        call.resolve()
    }
    
    @objc func withResponseHeaderFields(_ call: CAPPluginCall) {
        let tracker = call.getString("http_tracker")
        let headers = call.dictionaryWithValues(forKeys: ["http_headers"]) as NSDictionary
        if tracker != nil {
            implementation.withResponseHeaderFields(tracker_key: tracker!, responseHeaders: headers)
        }
        call.resolve()
    }
    
    @objc func withRequestHeaderFields(_ call: CAPPluginCall) {
        let tracker = call.getString("http_tracker")
        let headers = call.dictionaryWithValues(forKeys: ["http_headers"]) as NSDictionary
        if tracker != nil {
            implementation.withRequestHeaderFields(tracker_key: tracker!, requestHeaders: headers)
        }
        call.resolve()
    }
    
    @objc func withInstrumentationSource(_ call: CAPPluginCall) {
        let tracker = call.getString("http_tracker")
        let source = call.getString("information_source") ?? ""
        if tracker != nil {
            implementation.withInstrumentationSource(tracker_key: tracker!, informationSource: source)
        }
        call.resolve()
    }
    
    @objc func withErrorMessage(_ call: CAPPluginCall) {
        let tracker = call.getString("http_tracker")
        let error_message = call.getString("error_message") ?? ""
        if tracker != nil {
            implementation.withErrorMessage(tracker_key: tracker!, errorMessage: error_message)
        }
        call.resolve()
    }
    
    @objc func getCorrelationHeaders(_ call: CAPPluginCall) {
        let headers = implementation.getCorrelationHeaders()
        call.resolve([
            "headers": headers //NSDictionary
        ])
        return
    }
    
    @objc func startNextSession(_ call: CAPPluginCall) {
        implementation.startNextSession()
        call.resolve()
    }
    
    @objc func unblockScreenshots(_ call: CAPPluginCall) {
        implementation.unblockScreenshots()
        call.resolve()
    }
    
    @objc func blockScreenshots(_ call: CAPPluginCall) {
        implementation.blockScreenshots()
        call.resolve()
    }
    
    @objc func screenshotsBlocked(_ call: CAPPluginCall) {
        let is_blocked = implementation.screenshotsBlocked()
        call.resolve([
            "screenshots_blocked": is_blocked
        ])
        return
    }
    
    @objc func startSessionFrame(_ call: CAPPluginCall) {
        let sessionFrameName = call.getString("session_frame_name") ?? nil
        if sessionFrameName == nil {
            //do nothing and return nothing
            call.resolve()
            return
        }
        let session = implementation.startSessionFrame(sessionFrameName: sessionFrameName!)
        call.resolve([
            "session_frame": session
        ])
        return
    }
    
    @objc func endSessionFrame(_ call: CAPPluginCall) {
        let tracker = call.getString("session_frame")
        if tracker != nil {
            implementation.endSessionFrame(session_key: tracker!)
        }
        call.resolve()
    }
    
    @objc func updateSessionFrameName(_ call: CAPPluginCall) {
        let sessionFrameName = call.getString("session_frame_name") ?? nil
        let tracker = call.getString("session_frame")
        if sessionFrameName == nil || tracker == nil{
            //do nothing and return nothing
            call.resolve()
            return
        }
        //if here we should have vals for both and can force unwrap
        implementation.updateSessionFrameName(session_key: tracker!, session_name: sessionFrameName!)
        call.resolve([
            "session_frame": tracker!
        ])
        return
        
    }
    
    @objc func getVersion(_ call: CAPPluginCall) {
        call.resolve([
            "version": implementation.getVersion()
        ])
        return
    }
    
    @objc func clear(_ call: CAPPluginCall) {
        implementation.clear()
        call.resolve()
        return
    }
    
    override public func load() {
        if let appKey = getConfigValue("ADEUM_APP_KEY") as? String{
            config.appKey = appKey
        }
        if let collectorUrl = getConfigValue("ADEUM_COLLECTOR_URL") as? String{
            config.collectorURL = collectorUrl
        }
        if let screenshotUrl = getConfigValue("ADEUM_SCREENSHOT_URL") as? String{
            config.screenshotURL = screenshotUrl
        }
        if let screenshotEnabled = getConfigValue("ADEUM_SCREENSHOTS_ENABLED") as? Bool{
            config.screenshotsEnabled = screenshotEnabled
        }
        if let loggingLevel = getConfigValue("ADEUM_LOGGING_LEVEL") as? UInt{
            config.loggingLevel = ADEumLoggingLevel(rawValue: loggingLevel) ?? ADEUMInstrumentation.ADEumLoggingLevel.error
        }
        if let reachabilityHost = getConfigValue("ADEUM_REACHABILITY_HOST") as? String{
            config.reachabilityHostName = reachabilityHost
        }
        if let captureMode = getConfigValue("ADEUM_INTERACTION_CAPTURE_MODE") as? UInt{
            //let d = ADEumInteractionCaptureMode.ADEumInteractionCaptureModeNone
            config.interactionCaptureMode = ADEumInteractionCaptureMode.init(rawValue: captureMode)
        }
        implementation.pluginInitialize(config: config)
    }
}

File: android/src/main/java/com/appdynamics/appd/mrum/plugins/ionic/ADEUMMobileCapacitorPluginPlugin.java
package com.appdynamics.appd.mrum.plugins.ionic;

//import android.app.Instrumentation;
import android.util.Log;

import com.appdynamics.eumagent.runtime.AgentConfiguration;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;


@CapacitorPlugin(name = "ADEUMMobileCapacitorPlugin")
public class ADEUMMobileCapacitorPluginPlugin extends Plugin {
    private ADEUMMobileCapacitorPlugin implementation = new ADEUMMobileCapacitorPlugin();
    private static final String PLACEHOLDER_METRIC = "Generic Metric";
    private static final String PLACEHOLDER_KEY = "Generic Key";
//    private static final String PLACEHOLDER_VALUE = "Generic Value";
    private static final String TAG = "ADEUMCapacitorPlugin";

    @Override
    public void load() {
        String appKey = getConfig().getString("ADEUM_APP_KEY");
        //int collectorUrlResId = cordova.getActivity().getResources().getIdentifier("adeum_collector_url", "string", cordova.getActivity().getPackageName());
        String collectorUrl = getConfig().getString("ADEUM_COLLECTOR_URL");
        //int screenshotUrlResId = cordova.getActivity().getResources().getIdentifier("adeum_screenshot_url", "string", cordova.getActivity().getPackageName());
        String screenshotUrl = getConfig().getString("ADEUM_SCREENSHOT_URL");
        //int screenshotsEnabledResId = cordova.getActivity().getResources().getIdentifier("adeum_screenshots_enabled", "string", cordova.getActivity().getPackageName());
        boolean screenshotsEnabled = false;
        try {
            screenshotsEnabled = Boolean.parseBoolean(getConfig().getString("ADEUM_SCREENSHOTS_ENABLED"));
        } catch (Exception ignored) {
        }
        //int loggingLevelResId = cordova.getActivity().getResources().getIdentifier("adeum_logging_level", "string", cordova.getActivity().getPackageName());
        int loggingLevel = 2;
        try {
            loggingLevel = getConfig().getInt("ADEUM_LOGGING_LEVEL",0);

        } catch (Exception ignored) {
        }
        //int interactionCaptureModeResId = cordova.getActivity().getResources().getIdentifier("adeum_interaction_capture_mode", "string", cordova.getActivity().getPackageName());
        int interactionCaptureMode = 0;
        try {
            interactionCaptureMode = Integer.parseInt(getConfig().getString("ADEUM_INTERACTION_CAPTURE_MODE"));
        } catch (Exception ignored) {
        }

        AgentConfiguration config = AgentConfiguration.builder().
                withAppKey(appKey).
                withContext(this.getContext()).
                withCollectorURL(collectorUrl).
                withScreenshotURL(screenshotUrl).
                withLoggingLevel(loggingLevel).
                withJSAgentAjaxEnabled(true).
                withScreenshotsEnabled(screenshotsEnabled).
                withInteractionCaptureMode(interactionCaptureMode).
                build();

        // since there is no way to return an error to the host app we keep track of the error
        // in pluginInitialized flag and return it when a method is invoked on the plugin
        try {
            implementation.pluginInitialized(config);
        } catch (IllegalArgumentException ex) {
            Log.e(TAG, ex.getMessage());
            implementation.pluginInitialized = false;
        }
    }

    @PluginMethod
    public void echo(PluginCall call) {
        String value = call.getString("value");
        JSObject ret = new JSObject();
        ret.put("value", implementation.echo(value));
        call.resolve(ret);
    }

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public void startTimerWithName(PluginCall call) {
        String name = call.getString("name");
        Log.d(TAG, " startTimerWithName: " + name);
        implementation.startTimeWithName(name);
        call.resolve();
    }

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public void stopTimerWithName(PluginCall call) {
        String name = call.getString("name");
        name = name.replaceAll("\\s","").length()==0 ? PLACEHOLDER_METRIC : name;
        Log.d(TAG, " stopTimerWithName: " + name);
        implementation.stopTimerWithName(name);
        call.resolve();
    }

    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public void reportMetricWithName(PluginCall call) {
        String name = call.getString("name");
        name = name.replaceAll("\\s","").length()==0 ? PLACEHOLDER_METRIC : name;
        Log.d(TAG, " reportMetricWithName: " + name);
        Integer value = call.getInt("value");
        implementation.reportMetricWithName(name, value);
        call.resolve();
    }

    @PluginMethod
    public void leaveBreadcrumb(PluginCall call) {
        String name = call.getString("name");
        name = name.replaceAll("\\s","").length()==0 ? PLACEHOLDER_METRIC : name;
        Log.d(TAG, " leaveBreadcrumb: " + name);
        implementation.leaveBreadcrumb(name);
        call.resolve();
    }

    @PluginMethod
    public void setUserData(PluginCall call) {
        String key = call.getString("key");
        key = key.replaceAll("\\s","").length()==0 ? PLACEHOLDER_METRIC : key;
        String value = call.getString("value");
        value = value.replaceAll("\\s","").length()==0 ? PLACEHOLDER_METRIC : value;
        Log.d(TAG, " setUserData: " + key + ", " +value);
        if (!key.equals(PLACEHOLDER_KEY)) {
            implementation.setUserData(key, value);
        }
        call.resolve();
    }

    @PluginMethod
    public void removeUserData(PluginCall call){
        String key = call.getString("key");
        key = key.replaceAll("\\s","").length()==0 ? PLACEHOLDER_METRIC : key;
        Log.d(TAG, " removeUserData: " + key);
        if (!key.equals(PLACEHOLDER_KEY)) {
            implementation.removeUserData(key);
        }
        call.resolve();
    }

    @PluginMethod
    public void takeScreenshot(PluginCall call){
        implementation.takeScreenshot();
        call.resolve();
    }

    @PluginMethod
    public void beginCall(PluginCall call){
        String className = call.getString("className");
        String methodName = call.getString("methodName");
        JSArray args = call.getArray("withArguments");
        Log.d(TAG, " beginCall: " + className + methodName + args);
        if (className.replaceAll("\\s","").length()!=0 && methodName.replaceAll("\\s","").length()!=0){
            JSObject jsonObject = new JSObject();
            jsonObject.put("call_tracker", implementation.beginCall(className,methodName,args));
            call.resolve(jsonObject);
        }else{
            call.resolve();
        }
    }

    @PluginMethod
    public void endCall(PluginCall call){
        JSObject callTracker = call.getObject("call_tracker");
        if (callTracker != null) {
            implementation.endCall(callTracker);
        }
        call.resolve();
    }

    @PluginMethod
    public void beginHTTPRequest(PluginCall call){
        String url = call.getString("url");
        url = url.replaceAll("\\s","").length()==0 ? "" : url;
        Log.d(TAG, " beginHTTPRequest: " + url);
        if (url.length() <= 0){
            call.resolve();
            return;
        }
        String http_tracker = implementation.beginHttpRequest(url);
        JSObject ret = new JSObject();
        ret.put("http_tracker", http_tracker);
        call.resolve(ret);

    }

    @PluginMethod
    public void withURL(PluginCall call){
        //unimplemented due to ambiguity
    }

    @PluginMethod
    public void reportDone(PluginCall call){
        String tracker = call.getString("http_tracker");
        tracker = tracker.replaceAll("\\s","").length()==0 ? "" : tracker;
        Log.d(TAG, " reportDone: " + tracker);
        if (tracker.length() > 0){
            implementation.reportDone(tracker);
        }
        call.resolve();
    }

    @PluginMethod
    public void withResponseCode(PluginCall call){
        String tracker = call.getString("http_tracker");
        Integer statusCode = call.getInt("status_code");
        Log.d(TAG, " withResponseCode: " + tracker + ", " + statusCode);
        if (tracker.replaceAll("\\s","").length() > 0){
            implementation.withResponseCode(tracker, statusCode);
        }
        call.resolve();
    }

    @PluginMethod
    public void withResponseContentLength(PluginCall call){
        String tracker = call.getString("http_tracker");
        Integer contentLength = call.getInt("content_length");
        Log.d(TAG, " withResponseContentLength: " + tracker + ", " + contentLength);
        if (tracker.replaceAll("\\s","").length() > 0){
            implementation.withResponseContentLength(tracker, contentLength);
        }
        call.resolve();
    }

    @PluginMethod
    public void withRequestContentLength(PluginCall call){
        String tracker = call.getString("http_tracker");
        Integer contentLength = call.getInt("content_length");
        Log.d(TAG, " withResponseContentLength: " + tracker + ", " + contentLength);
        if (tracker.replaceAll("\\s","").length() > 0){
            implementation.withRequestContentLength(tracker, contentLength);
        }
        call.resolve();
    }

    @PluginMethod
    public void withResponseHeaderFields(PluginCall call){
        String tracker = call.getString("http_tracker");
        JSObject httpHeaders =  call.getObject("http_headers");
        if (tracker.replaceAll("\\s","").length() > 0){
            implementation.withResponseHeaderFields(tracker, httpHeaders);
        }
        call.resolve();
    }

    @PluginMethod
    public void withRequestHeaderFields(PluginCall call){
        String tracker = call.getString("http_tracker");
        JSObject httpHeaders =  call.getObject("http_headers");
        if (tracker.replaceAll("\\s","").length() > 0){
            implementation.withRequestHeaderFields(tracker, httpHeaders);
        }
        call.resolve();
    }

    @PluginMethod
    public void withErrorMessage(PluginCall call){
        String tracker = call.getString("http_tracker");
        String errorMessage = call.getString("error_message");
        if (tracker.replaceAll("\\s","").length() > 0){
            implementation.withErrorMessage(tracker, errorMessage);
        }
        call.resolve();
    }

    @PluginMethod
    public void getCorrelationHeaders(PluginCall call){
       JSObject correlationHeaders =  new JSObject();
       correlationHeaders.put("headers", implementation.getCorrelationHeaders());
       call.resolve(correlationHeaders);
    }

    @PluginMethod
    public void startNextSession(PluginCall call){
        implementation.startNextSession();
        call.resolve();
    }

    @PluginMethod
    public void unblockScreenshots(PluginCall call){
        implementation.unblockScreenshots();
        call.resolve();
    }

    @PluginMethod
    public void blockScreenshots(PluginCall call){
        implementation.blockScreenshots();
        call.resolve();
    }

    @PluginMethod
    public void screenshotsBlocked(PluginCall call){
        Boolean isBlocked = implementation.screenshotsBlocked();
        JSObject object = new JSObject();
        object.put("screenshots_blocked", isBlocked);
        call.resolve(object);
    }

    @PluginMethod
    public void startSessionFrame(PluginCall call){
        String sessionFrameName = call.getString("session_frame_name");
        if (sessionFrameName.replaceAll("\\s","").length() > 0){
            JSObject callObject = new JSObject();
            callObject.put("session_frame",  implementation.startSessionFrame(sessionFrameName));
            call.resolve(callObject);
        }else{
            call.resolve();
        }
    }

    @PluginMethod
    public void endSessionFrame(PluginCall call){
        String sessionFrame = call.getString("session_frame");
        if (sessionFrame.replaceAll("\\s","").length() > 0){
            implementation.endSessionFrame(sessionFrame);
        }
        call.resolve();
    }

    @PluginMethod
    public void updateSessionFrame(PluginCall call){
        String sessionFrameName = call.getString("session_frame_name");
        String tracker = call.getString("session_frame");
        if (sessionFrameName.replaceAll("\\s","").length() > 0
                && tracker.replaceAll("\\s","").length() > 0) {
            implementation.updateSessionFrame(tracker, sessionFrameName);
            JSObject object = new JSObject();
            object.put("session_frame", tracker);
            call.resolve(object);
        }else{
            call.resolve();
        }
    }



    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public void clear(PluginCall call) {
        implementation.clear();
    }

    @PluginMethod
    public void checkPluginInitialized(PluginCall call) {
        boolean is_initialized = false;
        is_initialized = implementation.checkPluginInitialized();
        //JSObject ret = new JSObject();
        if (!is_initialized) {
            //this may be bad and should just return true or false or reject and resolve
            call.reject("Plugin was not initialized due to missing or invalid App Key.");
            return;
        } else {
            //ret.put("initialized", is_initialized);
            //call.resolve(ret);
            call.resolve();
        }
    }


    @PluginMethod
    public void crash(PluginCall call) {
        JSObject ret = new JSObject();
        try {
            implementation.crash();
        } catch (RuntimeException re) {
            ret.put("exception_type_name", re.getClass().getCanonicalName());
            ret.put("exception_message", re.getMessage());
            call.resolve(ret);
        }

    }
}
